ArgsParsing/ArgsParseResult.cs

```diff
@@ -0,0 +1,166 @@
+﻿using System.Collections.Immutable;
+
+namespace ArgsParsing
+{
+    public enum ErrorRelevanceConfidence
+    {
+        /// <summary>
+        /// Indicates that this error is unlikely to be helpful in determining why overall parsing might have failed.
+        /// Other errors should preferably be reported instead.
+        /// </summary>
+        Unlikely,
+
+        /// <summary>
+        /// Indicates that this error does not know or does not want to express its meaningfulness.
+        /// </summary>
+        Default,
+
+        /// <summary>
+        /// Indicates that this error is likely to accurately describe why parsing failed overall,
+        /// and should preferably be reported over other, less helpful errors.
+        /// </summary>
+        Likely,
+    }
+
+    public struct Success<T>
+    {
+        public T Result { get; }
+        public IImmutableList<string> RemainingArgs { get; }
+
+        public Success(T result, IImmutableList<string> remainingArgs)
+        {
+            Result = result;
+            RemainingArgs = remainingArgs;
+        }
+    }
+
+    public struct Failure
+    {
+        public ErrorRelevanceConfidence Relevance { get; }
+        public string Error { get; }
+
+        public Failure(ErrorRelevanceConfidence relevance, string error)
+        {
+            Relevance = relevance;
+            Error = error;
+        }
+    }
+
+    /// <summary>
+    /// A result object that can either be successful or not,
+    /// and contains an instance of <typeparamref name="T"/> on success.
+    /// </summary>
+    /// <typeparam name="T">type of the success object that will be contained on success.</typeparam>
+    public struct ArgsParseResult<T>
+    {
+        /// <summary>
+        /// Whether this success object resulted from a successful parse operation, and therefore contains a result.
+        /// </summary>
+        public bool IsSuccess { get; }
+        /// <summary>
+        /// The result object, if successful.
+        /// </summary>
+        public Success<T> SuccessResult { get; }
```

Comment:
I've got an better idea, let me throw an exception if someone tries to access this with `IsSuccess` returning `false`. Because callers should need to check for `IsSuccess`, not for nullability on the actual result